datafusion/common/src/scalar/mod.rs

```diff
@@ -1177,40 +1185,103 @@ impl ScalarValue {
             | ScalarValue::Float64(None) => Ok(self.clone()),
             ScalarValue::Float64(Some(v)) => Ok(ScalarValue::Float64(Some(-v))),
             ScalarValue::Float32(Some(v)) => Ok(ScalarValue::Float32(Some(-v))),
-            ScalarValue::Int8(Some(v)) => Ok(ScalarValue::Int8(Some(-v))),
-            ScalarValue::Int16(Some(v)) => Ok(ScalarValue::Int16(Some(-v))),
-            ScalarValue::Int32(Some(v)) => Ok(ScalarValue::Int32(Some(-v))),
-            ScalarValue::Int64(Some(v)) => Ok(ScalarValue::Int64(Some(-v))),
+            ScalarValue::Int8(Some(v)) => Ok(ScalarValue::Int8(Some(v.neg_checked()?))),
+            ScalarValue::Int16(Some(v)) => Ok(ScalarValue::Int16(Some(v.neg_checked()?))),
+            ScalarValue::Int32(Some(v)) => Ok(ScalarValue::Int32(Some(v.neg_checked()?))),
+            ScalarValue::Int64(Some(v)) => Ok(ScalarValue::Int64(Some(v.neg_checked()?))),
             ScalarValue::IntervalYearMonth(Some(v)) => {
-                Ok(ScalarValue::IntervalYearMonth(Some(-v)))
+                Ok(ScalarValue::IntervalYearMonth(Some(neg_checked_with_ctx(
+                    *v,
+                    format!("In negation of IntervalYearMonth({v})"),
+                )?)))
             }
             ScalarValue::IntervalDayTime(Some(v)) => {
                 let (days, ms) = IntervalDayTimeType::to_parts(*v);
-                let val = IntervalDayTimeType::make_value(-days, -ms);
+                let val = IntervalDayTimeType::make_value(
+                    neg_checked_with_ctx(
+                        days,
+                        format!("In negation of days {days} in IntervalDayTime"),
```

Comment:
I think that calling these functions like this will  result in actually creating these error strings for each row, even if there is no error -- I think it is important to avoid that overhead.

I suspect it would be faster to use `map_err` like

```rust
let days = v.neg_checked()
  .map_err(|e| {
    DataFusionError::from(e)
      .context(format!("In negation of milliseconds {ms} in IntervalDayTime")))
  })?
```

The reason being that in that formulation the `format!` is part of a closure that is only called when there is an actual error

Reply:
Yeah, I first tried that, but I didn't like repeating the same thing for every case. But this went over my head, I'll fix this back to `map_err`. Thanks for the tip!